---
Title: "Intro to Categorical Model-Making Module by Austin Hayes"

Author:
  - Name: Austin Hayes

  -  Email: ahayes65@charlotte.edu

  -  Affiliation: University of North Carolina at Charlotte

Date: July 29, 2025

Description: Using WNBA Team 'Per Game' stats, we will review the basics of categorical model-making. We will make QDA, LDA and Logistic models in order to predict which team will become the next WNBA Champion.

Categories:
  - Logistic Regression
  - Quadratic Discriminant Analysis
  - Linear Discriminant Analysis
  - Importing and Reading data
  - Data Cleaning
  - Data Science
  - Pandas


### Data

This Dataset is from Basketball Reference @ https://www.basketball-reference.com

Visit the original data page here: https://www.basketball-reference.com/wnba/years/2024.html

The data set contains 60 rows and 26 columns. Each row represents a WNBA team during the 2024 season.

Download data: 

Available on the [Intro to Categorical Model-Making Module by Austin Hayes](https://github.com/schuckers/Charlotte_SCORE_Summer25/tree/main/Data%20for%20Modules/Data%20For%20Intro%20to%20Categorical%20Model-Making%20Module%20by%20Austin%20Hayes): [2024_WNBA_Per_Game.csv](https://raw.githubusercontent.com/schuckers/Charlotte_SCORE_Summer25/refs/heads/main/Data%20for%20Modules/Data%20For%20Intro%20to%20Categorical%20Model-Making%20Module%20by%20Austin%20Hayes/2024_WNBA_Per_Game.csv)

---

### Variables and their Descriptions:


<details>
<summary><b>Variable Descriptions</b></summary>

| Variable | Description |
|----------|-------------|
| Rk       | Rank of the team in the league |
| Team     | Name of the team |
| G        | Games played |
| MP       | Minutes played per game |
| FG       | Field Goals made per game |
| FGA      | Field Goal attempts per game |
| FG%      | Field Goal percentage (FG ÷ FGA) |
| 3P       | Three-Point Field Goals made per game |
| 3PA      | Three-Point Field Goal attempts per game |
| 3P%      | Three-Point Field Goal percentage (3P ÷ 3PA) |
| 2P       | Two-Point Field Goals made per game |
| 2PA      | Two-Point Field Goal attempts per game |
| 2P%      | Two-Point Field Goal percentage (2P ÷ 2PA) |
| FT       | Free Throws made per game |
| FTA      | Free Throw attempts per game |
| FT%      | Free Throw percentage (FT ÷ FTA) |
| ORB      | Offensive Rebounds per game |
| DRB      | Defensive Rebounds per game |
| TRB      | Total Rebounds per game |
| AST      | Assists per game |
| STL      | Steals per game |
| BLK      | Blocks per game |
| TOV      | Turnovers per game |
| PF       | Personal Fouls per game |
| PTS      | Points scored per game |
| CHAMPION | Did that team win the championship that year? 1 = Yes, 0 = No |

</details>

---

# Review Questions

Below, you will be asked a series of questions to review the material you have learned throughout this module.

SETUP:

In [1]:
# Import the necessary librarys for the module
# Basic Data Science Library for importing data and data manipulation
import pandas as pd

# Library for mathematical operations
import numpy as np

# Import only the necessary functions from the sklearn library
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix, accuracy_score, precision_score, recall_score, f1_score
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis, QuadraticDiscriminantAnalysis
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import StandardScaler

In [2]:
# Read in the WNBA 2020-2024 data from Github 
# This dataset will be used to train the final model
TRAIN_WNBA_20_24_Data = pd.read_csv('https://raw.githubusercontent.com/schuckers/Charlotte_SCORE_Summer25/refs/heads/main/Data%20for%20Modules/Data%20For%20Intro%20to%20Categorical%20Model-Making%20Module%20by%20Austin%20Hayes/WNBA_PER_GAME_20_24_DATA.csv')

In [3]:
# Read in the current WNBA 2025 stats for our model prediction later on in this module
CURRENT_WNBA_2025_Data = pd.read_csv('https://raw.githubusercontent.com/schuckers/Charlotte_SCORE_Summer25/refs/heads/main/Data%20for%20Modules/Data%20For%20Intro%20to%20Categorical%20Model-Making%20Module%20by%20Austin%20Hayes/2025_WNBA_Per_Game.csv')

### 1.) How does KNN work and what needs to be done to the data before use of a KNN model?

ANSWER: _KNN classifys points based on their "nearest neighbors". This means the model compares the new points to similar points that the model was trained on in order to make a prediction. K is the number of neighbors that are used to classify a new point. K should also be an odd number in order to prevent a tie among K points. Before using KNN, it is important to scale your data so that the model does not make biased predictions when using predictors with larger numerical ranges._

### 2.) Using the 2020-2024 dataset, create a list of predictor variables that include FG%, 3P%, 2P%, FT%, ORB, AST. Scale the list of predictor variables. Make sure you also scale them for the 2025 data since you will be asked to make a prediction on them later. Create a new column in the ORIGINAL 2020-2024 dataset that assesses whether a teams points ('PTS') are over 82 per game. 1 = True, 0 = False. Make this column for the ORIGINAL 2025 dataset as well. 

Hint: _Use .astype(int)_ 

In [4]:
# ANSWER:

# Create the scaler object
scaler = StandardScaler()

# Fit the scaler to the 2020-2024 WNBA data and transform (scale) the data
training_data_20_24 = scaler.fit_transform(TRAIN_WNBA_20_24_Data[['FG%','3P%','FT%','ORB','AST']])

# Transform (scale) the 2025 WNBA data using the same scaler
test_data_2025 = scaler.fit_transform(CURRENT_WNBA_2025_Data[['FG%','3P%','FT%','ORB','AST']])

# Create a new column in the 2020-2024 dataset that assesses whether a team's points ('PTS') are over 82 per game
TRAIN_WNBA_20_24_Data['OVR82?'] = (TRAIN_WNBA_20_24_Data['PTS'] > 82).astype(int)

# Make a new column for the 2025 data as well so we can record an accuracy score.
CURRENT_WNBA_2025_Data['OVR82?'] = (CURRENT_WNBA_2025_Data['PTS'] > 82).astype(int)

In [5]:
# Check to see that new column works and is established as 1's and 0's
TRAIN_WNBA_20_24_Data

Rk                     Team   G     MP    FG   FGA    FG%    3P   3PA  \
0    1      2024 Las Vegas Aces  40  200.6  30.9  68.1  0.454   9.4  26.5   
1    2    2024 New York Liberty  40  200.0  30.8  68.7  0.448  10.1  29.0   
2    3       2024 Indiana Fever  40  200.6  31.3  68.5  0.456   9.2  25.9   
3    4        2024 Dallas Wings  40  201.9  31.7  71.0  0.446   6.3  19.2   
4    5       2024 Seattle Storm  40  201.2  31.1  71.3  0.435   6.1  21.0   
5    6      2024 Minnesota Lynx  40  201.9  30.1  67.3  0.448   9.5  25.0   
6    7     2024 Phoenix Mercury  40  201.2  29.1  66.3  0.439   8.5  26.2   
7    8     2024 Connecticut Sun  40  201.2  29.3  65.9  0.444   5.9  18.0   
8    9  2024 Washington Mystics  40  201.2  29.0  67.0  0.433   9.7  26.6   
9   10  2024 Los Angeles Sparks  40  200.6  28.1  66.4  0.423   7.2  22.6   
10  11         2024 Chicago Sky  40  200.0  29.7  70.3  0.422   4.8  14.9   
11  12       2024 Atlanta Dream  40  201.9  27.8  68.1  0.408   6.0  19.4   
12   1      2023 Las Vegas Aces  40  200.0  33.6  69.2  0.486   9.3  24.9   
13   2    2023 New York Liberty  40  201.9  32.2  69.9  0.460  11.1  29.7   
14   3       2023  Dallas Wings  40  200.6  32.4  73.2  0.443   6.8  21.3   
15   4     2023 Connecticut Sun  40  201.9  30.2  67.8  0.445   7.2  20.0   
16   5       2023 Atlanta Dream  40  201.3  29.4  68.7  0.428   6.4  19.2   
17   6         2023 Chicago Sky  40  201.3  30.7  69.6  0.442   8.3  22.2   
18   7       2023 Indiana Fever  40  201.9  30.2  68.3  0.442   6.7  19.8   
19   8  2023 Washington Mystics  40  200.6  28.9  67.5  0.428   7.8  23.1   
20   9      2023 Minnesota Lynx  40  201.3  29.4  67.5  0.435   6.8  20.8   
21  10  2023 Los Angeles Sparks  40  200.6  28.4  66.8  0.425   6.5  19.5   
22  11       2023 Seattle Storm  40  200.6  27.9  67.9  0.410   8.5  24.2   
23  12     2023 Phoenix Mercury  40  200.6  28.4  63.9  0.444   6.8  20.7   
24   1      2022 Las Vegas Aces  36  200.7  32.3  70.3  0.460   9.5  26.4   
25   2         2022 Chicago Sky  36  202.1  32.7  67.9  0.481   7.2  20.9   
26   3     2022 Connecticut Sun  36  201.4  31.9  69.0  0.462   6.4  18.0   
27   4        2022 Dallas Wings  36  200.7  29.7  68.2  0.435   7.8  22.6   
28   5       2022 Seattle Storm  36  200.7  30.4  68.8  0.442   9.3  25.6   
29   6      2022 Minnesota Lynx  36  201.4  30.2  67.1  0.450   6.8  19.6   
30   7     2022 Phoenix Mercury  36  202.1  28.6  66.5  0.429   8.8  26.7   
31   8  2022 Washington Mystics  36  202.1  29.3  66.8  0.439   7.6  22.5   
32   9    2022 New York Liberty  36  202.1  28.8  66.8  0.431   9.7  27.6   
33  10  2022 Los Angeles Sparks  36  200.7  29.8  66.9  0.446   5.4  16.9   
34  11       2022 Atlanta Dream  36  201.4  28.8  68.4  0.420   7.5  21.4   
35  12       2022 Indiana Fever  36  201.4  28.4  69.4  0.409   6.8  20.1   
36   1      2021 Las Vegas Aces  32  202.3  33.1  70.2  0.472   5.1  13.5   
37   2       2021 Seattle Storm  32  203.9  31.0  68.9  0.450   8.6  22.6   
38   3         2021 Chicago Sky  32  203.1  30.8  69.9  0.441   7.3  21.2   
39   4      2021 Minnesota Lynx  32  201.6  30.7  67.1  0.458   6.8  19.8   
40   5     2021 Phoenix Mercury  32  201.6  29.5  65.7  0.450   7.5  21.8   
41   6        2021 Dallas Wings  32  201.6  29.7  70.7  0.420   8.2  23.1   
42   7  2021 Washington Mystics  32  200.8  28.4  69.2  0.410   9.0  27.4   
43   8     2021 Connecticut Sun  32  201.6  29.2  65.6  0.444   7.0  19.6   
44   9       2021 Atlanta Dream  32  201.6  30.6  73.5  0.417   6.1  19.8   
45  10    2021 New York Liberty  32  200.8  28.3  66.1  0.428  10.0  27.7   
46  11       2021 Indiana Fever  32  200.0  28.4  68.2  0.417   4.9  17.3   
47  12  2021 Los Angeles Sparks  32  201.6  27.0  65.8  0.411   6.8  20.1   
48   1      2020 Las Vegas Aces  22  200.0  32.7  68.7  0.477   4.2  11.5   
49   2       2020 Seattle Storm  22  200.0  31.9  67.8  0.470   8.4  21.4   
50   3         2020 Chicago Sky  22  200.0  33.5  68.2  0.491   7.6  21.6   


### 3.) Using the scaled data you created in question 2, predict which team(s) will have above 82 points per game during the 2025 WNBA season with a QDA model. Record an accuracy score.

QUICK NOTE: You can use an accuracy score in this case because there are already pts per game recorded for each team thus far. Although they are not final for the WHOLE season, this is an interesting experiment to see if these are strong predictors of whether a team will have high points per game. 

In [6]:
# Create the model object
qda_model = QuadraticDiscriminantAnalysis()

# Fit the model using the 2020-2024 WNBA data that was scaled earlier
qda_model.fit(training_data_20_24, TRAIN_WNBA_20_24_Data['OVR82?'])

# Have the model make predictions using the 2025 WNBA data
qda_predictions = qda_model.predict(test_data_2025)

# takes a look at the array of predictions
qda_predictions

array([1, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0])

In [7]:
# Record the accuracy score of the QDA model
qda_accuracy = accuracy_score(CURRENT_WNBA_2025_Data['OVR82?'], qda_predictions)

In [8]:
# Print the accuracy score
print(f"QDA Model Accuracy: {qda_accuracy:.2f}")

QDA Model Accuracy: 0.85


### 4.) Do the same thing you did in question 3 but with 3 KNN models instead. Use K=5, K=25 and K=51 respectively. Record an accuracy score.

ANSWER FOR K=5:

In [9]:
# Create the model object
knn_model_5 = KNeighborsClassifier(n_neighbors=5)

# Fit the model using the 2020-2024 WNBA data that was scaled earlier
knn_model_5.fit(training_data_20_24, TRAIN_WNBA_20_24_Data['OVR82?'])

# Have the model make predictions using the 2025 WNBA data
knn_predictions_5 = knn_model_5.predict(test_data_2025)

# takes a look at the array of predictions
knn_predictions_5

array([1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0])

In [10]:
# Record the accuracy score of the QDA model
knn_accuracy_5 = accuracy_score(CURRENT_WNBA_2025_Data['OVR82?'], knn_predictions_5)

# Print the accuracy score
print(f"KNN Model Accuracy for 5 neighbors: {knn_accuracy_5:.2f}")

KNN Model Accuracy for 5 neighbors: 0.85


ANSWER FOR K=25:

In [11]:
# Create the model object
knn_model_25 = KNeighborsClassifier(n_neighbors=25)

# Fit the model using the 2020-2024 WNBA data that was scaled earlier
knn_model_25.fit(training_data_20_24, TRAIN_WNBA_20_24_Data['OVR82?'])

# Have the model make predictions using the 2025 WNBA data
knn_predictions_25 = knn_model_25.predict(test_data_2025)

# takes a look at the array of predictions
knn_predictions_25

array([1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [12]:
# Record the accuracy score of the QDA model
knn_accuracy_25 = accuracy_score(CURRENT_WNBA_2025_Data['OVR82?'], knn_predictions_25)

# Print the accuracy score
print(f"KNN Model Accuracy for 25 neighbors: {knn_accuracy_25:.2f}")

KNN Model Accuracy for 25 neighbors: 0.77


ANSWER FOR K=51

In [13]:
# Create the model object
knn_model_51 = KNeighborsClassifier(n_neighbors=51)

# Fit the model using the 2020-2024 WNBA data that was scaled earlier
knn_model_51.fit(training_data_20_24, TRAIN_WNBA_20_24_Data['OVR82?'])

# Have the model make predictions using the 2025 WNBA data
knn_predictions_51 = knn_model_51.predict(test_data_2025)

# takes a look at the array of predictions
knn_predictions_51

array([1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0])

In [14]:
# Record the accuracy score of the QDA model
knn_accuracy_51 = accuracy_score(CURRENT_WNBA_2025_Data['OVR82?'], knn_predictions_51)

# Print the accuracy score
print(f"KNN Model Accuracy for 51 neighbors: {knn_accuracy_51:.2f}")

KNN Model Accuracy for 51 neighbors: 0.54


### 5.) As you did in questions 3 and 4, make a prediction but use a Logistic model instead. Record an accuracy score.

In [15]:
# Create the model object
log_model = LogisticRegression()

# Fit the model using the 2020-2024 WNBA data that was scaled earlier
log_model.fit(training_data_20_24, TRAIN_WNBA_20_24_Data['OVR82?'])

# Have the model make predictions using the 2025 WNBA data
log_predictions = log_model.predict(test_data_2025)

# takes a look at the array of predictions
log_predictions

array([1, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0])

In [16]:
# Record the accuracy score of the QDA model
log_accuracy = accuracy_score(CURRENT_WNBA_2025_Data['OVR82?'], log_predictions)

# Print the accuracy score
print(f"Logistic Model Accuracy: {log_accuracy:.2f}")

Logistic Model Accuracy: 0.85


### 6.) Which model(s) had the highest accuracy score? Is there a tie between the most accurate models? Or is there a model that is clearly the best in terms of an accuracy score?

ANSWER: _The Logistic model, the QDA model, and the KNN model with K=5 are tied for the most accurate models with 85% accuracy!_

### 7.) What differentiates LDA and QDA? List a few differences.

ANSWER:

- LDA assumes the same variance between categorical groups.
- QDA is flexible and allows different variances.
- QDA allows for a non-linear, quadratic decision boundary between categorical groups
- LDA only allows for a linear decision boundary between categorical groups.